### PR TITLE
Reorganized signature-scheme detection in 'apps/s_cb.c:security_callback_debug' function

### PR DESCRIPTION
**Pull request by lzsiga**
Reorganized signature-scheme detection in 'apps/s_cb.c:security_callback_debug' callback-function.

So far, it only handled hash-and-algorithm pairs from TLS1.2,
now it also handles 'schemes' defined in TLS1.3 like 0x0807=ed25519 or
0x0809=rsa_pss_pss_sha256

Now it prints information in one of these formats:
```
... Algorithm scheme=ecdsa_secp256r1_sha256, security bits=128 ... TLS1.3
... Algorithm digest=SHA384, algorithm=DSA, security bits=192  ... TLS1.2
... Algorithm scheme=unknown(0x0e01), security bits=128        ... unhandled case
```
To implement this added three new lookup-tables: signature_tls13_scheme_list,
signature_tls12_alg_list, signature_tls12_hash_list.

Also minor changes in 'security_callback_debug', eg adding variable 'show_nm'
to indicate if we should show 'nm'.

##### Checklist

- [ ] documentation is added or updated
- [ ] tests are added or updated

### CODE REVIEW

**Comment by lzsiga** [comment]
Sorry, forgot to say: I'd like to mark this as 'trivial change'

File: apps/s_cb.c [hunk 1]
```diff
@@ -1373,27 +1428,26 @@ static int security_callback_debug(const SSL *s, const SSL_CTX *ctx,
...
-            BIO_puts(sdb->out, OBJ_nid2sn(nid));
-            if (sname)
-                BIO_printf(sdb->out, ", algorithm=%s", sname);
-            else
-                BIO_printf(sdb->out, ", algid=%d", salg[1]);
-            break;
+            BIO_printf(sdb->out, "%s", nm);
+
+            sname= lookup(raw_sig_code, signature_tls13_scheme_list, NULL);
+            if (sname) {
```
Comment (mattcaswell) [inline apps/s_cb.c]: Style nit: Make this `sname != NULL`

File: apps/s_cb.c [hunk 2]
```diff
@@ -1373,27 +1428,26 @@ static int security_callback_debug(const SSL *s, const SSL_CTX *ctx,
...
+                BIO_printf(sdb->out, " scheme=%s", sname);
+            } else {
+                int alg_code= salg[1];
+                int hash_code= salg[0];
+                const char *alg_str= lookup(alg_code, signature_tls12_alg_list, NULL);
+                const char *hash_str= lookup(hash_code, signature_tls12_hash_list, NULL);
+
+                if (alg_str && hash_str) {
```
Comment (mattcaswell) [inline apps/s_cb.c]: Style nit: Make this `alg_str != NULL && hash_str != NULL`

File: apps/s_cb.c [hunk 3]
```diff
@@ -1304,29 +1351,37 @@ static int security_callback_debug(const SSL *s, const SSL_CTX *ctx,
...
         break;
+    case SSL_SECOP_SIGALG_SUPPORTED:
+    case SSL_SECOP_SIGALG_SHARED:
+    case SSL_SECOP_SIGALG_CHECK:
+    case SSL_SECOP_SIGALG_MASK:
+        show_nm = 0;
+        break;
     }
-    if (nm != NULL)
+    if (show_nm)
```
Comment (mattcaswell) [inline apps/s_cb.c]: Why is this change necessary?

Comment (lzsiga) [reply apps/s_cb.c]: Hi, thank you for checking the code; in this case variable _'nm'_ does contain a literal I wish to print; problem is line 1385 (line 1330 in original) which prints an =equal sign after nm: 
```
-    if (nm != NULL)
+    if (show_nm)
         BIO_printf(sdb->out, "%s=", nm);
```
this doesn't fit into the new behaviour (it either prints 'scheme=xxx' or 'algorithm=xxx, hash=yyy'), so I had to find a way to suppress printing of 'nm' without destructing its value.

Another option would be calling `nm = lookup(op, callback_types, NULL);` once more in `case SSL_SECOP_OTHER_SIGALG`.

File: apps/s_cb.c (hunk 1 above)
Comment (lzsiga) [reply apps/s_cb.c]: fixed this (at-least in my fork), and the other coding style note. I'll try to send a printed/signed/scanned cla within 24 hours.

File: apps/s_cb.c (hunk 3 above)
Comment (lzsiga) [reply apps/s_cb.c]: Though 'nm' can still be NULL when I wish to print it, so I added one more check:
```
            if (nm != NULL) BIO_printf(sdb->out, "%s", nm);
            else            BIO_printf(sdb->out, "s_cb.c:security_callback_debug op=0x%x", op);
```

File: apps/s_cb.c [hunk 4]
```diff
@@ -1373,27 +1428,27 @@ static int security_callback_debug(const SSL *s, const SSL_CTX *ctx,
...
+                /* raw_sig_code: signature_scheme from tls1.3, or signature_and_hash from tls1.2 */
 
-            BIO_puts(sdb->out, OBJ_nid2sn(nid));
-            if (sname)
-                BIO_printf(sdb->out, ", algorithm=%s", sname);
-            else
-                BIO_printf(sdb->out, ", algid=%d", salg[1]);
-            break;
+            if (nm != NULL) BIO_printf(sdb->out, "%s", nm);
+            else            BIO_printf(sdb->out, "s_cb.c:security_callback_debug op=0x%x", op);
```
Comment (mattcaswell) [inline apps/s_cb.c]: Style nit: Should be like this

````
if (nm != NULL)
    xxx;
else
    yyy;
````
i.e. the the if and else branches need to be on separate lines

File: apps/s_cb.c [hunk 5]
```diff
@@ -1373,27 +1428,27 @@ static int security_callback_debug(const SSL *s, const SSL_CTX *ctx,
...
+
+            sname= lookup(raw_sig_code, signature_tls13_scheme_list, NULL);
+            if (sname != NULL) {
+                BIO_printf(sdb->out, " scheme=%s", sname);
+            } else {
+                int alg_code= salg[1];
+                int hash_code= salg[0];
+                const char *alg_str= lookup(alg_code, signature_tls12_alg_list, NULL);
+                const char *hash_str= lookup(hash_code, signature_tls12_hash_list, NULL);
```
Comment (mattcaswell) [inline apps/s_cb.c]: Style nit: Needs a space on left side of "=" on each of the above lines. And in fact also in assignment of raw_sig_code and sname above.

File: apps/s_cb.c [hunk 6]
```diff
@@ -1373,27 +1428,27 @@ static int security_callback_debug(const SSL *s, const SSL_CTX *ctx,
...
+
+                if (alg_str != NULL && hash_str != NULL) {
+                    BIO_printf(sdb->out, " digest=%s, algorithm=%s", hash_str, alg_str);
+                } else {
+                    BIO_printf(sdb->out, " scheme=unknown(0x%04x)", raw_sig_code);
+                }
```
Comment (mattcaswell) [inline apps/s_cb.c]: These are both single lines for both if and else, so no {} required

File: apps/s_cb.c (hunk 3 above)
Comment (mattcaswell) [reply apps/s_cb.c]: Ok

**Comment by mattcaswell** [comment]
Close/reopen to kick CLA bot

**Comment by mattcaswell** [comment]
Hmmm. CLA is registered but clabot has not recognised it. Looks like the author email in the commits is different to the CLA registered email address. Please can you update the commits to amend the author to use the registered email.

File: apps/s_cb.c (hunk 4 above)
Comment (lzsiga) [reply apps/s_cb.c]: Done

File: apps/s_cb.c (hunk 5 above)
Comment (lzsiga) [reply apps/s_cb.c]: Done

File: apps/s_cb.c (hunk 6 above)
Comment (lzsiga) [reply apps/s_cb.c]: done

**Comment by lzsiga** [comment]
I think by now I have committed every possible git-related errors. Would it be possible to delete this pull request (also my forked lzsiga/openssl) and start it from scratch?

**Comment by mattcaswell** [comment]
> I think by now I have committed every possible git-related errors. Would it be possible to delete this pull request (also my forked lzsiga/openssl) and start it from scratch?

If you want to you can. Although you can also just make whatever fixes you want to your commits (squash/rebase etc as required) and then just do a force push.

**Comment by lzsiga** [comment]
I admit, I don't know the first thing about git, so it's now wonder I've created chaos (commited changes with wrong email, also with wrong user etc). I'd like to close this now, and start from the beginning. (Of course including the fixes/corrections from matcaswell.)
